File: python/yolo8_image.py
import json

import cv2
import yaml
from wanderary import Box2DDrawer, ImageCropROIs, YOLOv8


def create_drawer(class_num, names_path):
    class_names = [""] * 80
    with open(names_path, "r") as f:
        names = json.load(f)
        for i, name in enumerate(names):
            class_names[int(i)] = name
    return Box2DDrawer(80, class_names)


if __name__ == '__main__':
    cfgpath = "configs/stat_yolov8_time_cfg.yaml"
    imgpath = "data/000000460347.jpg"
    typespath = "data/coco_type_names.json"
    savepath = ".cache/000000460347_pyres.jpg"

    # 加载配置/图像/可视化
    with open(cfgpath, 'r') as f:
        cfg = yaml.safe_load(f)
    img = cv2.imread(imgpath)
    drawer = create_drawer(80, typespath)

    # 初始化
    yolo = YOLOv8("yolov8", cfg, 1)

    # 推理
    rois = ImageCropROIs(img.shape[1::-1], (-1, -1), (-1, -1), False)
    bbox2ds = yolo.run(img, rois)

    # 可视化
    vis = drawer.draw(bbox2ds, img)
    cv2.imwrite(savepath, vis)

File: examples/yolo8_image.cpp
#include <string>
#include <vector>

#include <glog/logging.h>
#include <wanderary/apps/yolo.h>
#include <wanderary/utils/convertor.h>
#include <wanderary/utils/yaml_utils.h>
#include <wanderary/visualization/draw_boxes.h>

using YOLOv8 = wdr::apps::YOLOv8;
using Box2DDrawer = wdr::vis::Box2DDrawer;

Box2DDrawer CreateDrawer(const std::string &names_path) {
  const auto names = wdr::LoadJson(names_path);
  const int class_num = names.size();
  std::vector<std::string> class_names(class_num);
  for (const auto &item : names.items())
    class_names[std::stoi(item.key())] = item.value();
  return Box2DDrawer(class_num, class_names);
}

int main(int argc, char **argv) {
  // 数据加载
  const std::string cfgpath = "configs/stat_yolov8_time_cfg.yaml";
  const std::string imgpath = "data/000000460347.jpg";
  const std::string typespath = "data/coco_type_names.json";
  const std::string savepath = ".cache/000000460347_res.jpg";

  // 加载配置/图像/可视化
  const auto cfg = wdr::LoadYaml<wdr::json>(cfgpath);
  const auto img = cv::imread(imgpath);
  Box2DDrawer drawer = CreateDrawer(typespath);

  // Yolo初始化
  YOLOv8 yolov8("yolov8", cfg, 1);

  // 推理
  const auto rois =
      wdr::ImageCropROIs(img.size(), cv::Size(-1, -1), cv::Size(-1, -1), false);
  const auto box2ds = yolov8.run(img, rois, nullptr);

  // 可视化
  cv::Mat vis;
  img.copyTo(vis);
  drawer.draw(box2ds, &vis);
  cv::imwrite(savepath, vis);
  LOG(INFO) << "Save visualization to " << savepath;
  return 0;
}
